grpc/src/main/java/io/micrometer/core/instrument/binder/grpc/TaggingMetricCollectingServerInterceptor.java

```diff
@@ -0,0 +1,147 @@
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+
+/**
+ * Extension of the micrometer's {@link MetricCollectingServerInterceptor} that enables us to add
+ * extra tags on each {@link ServerCall}.
+ */
+public class TaggingMetricCollectingServerInterceptor extends MetricCollectingServerInterceptor {
+
+  /**
+   * The total number of requests received, same as
+   * MetricCollectingServerInterceptor#METRIC_NAME_SERVER_REQUESTS_RECEIVED.
+   */
+  private static final String METRIC_NAME_SERVER_REQUESTS_RECEIVED =
+      "grpc.server.requests.received";
+
+  /**
+   * The total number of responses sent, same as
+   * MetricCollectingServerInterceptor#METRIC_NAME_SERVER_RESPONSES_SENT.
+   */
+  private static final String METRIC_NAME_SERVER_RESPONSES_SENT = "grpc.server.responses.sent";
+
+  /**
+   * The total time taken for the server to complete the call, same as
+   * MetricCollectingServerInterceptor#METRIC_NAME_SERVER_PROCESSING_DURATION.
+   */
+  private static final String METRIC_NAME_SERVER_PROCESSING_DURATION =
+      "grpc.server.processing.duration";
+
+  /** Stargate {@link GrpcMetricsTagProvider} for extra tags on each call. */
+  private final GrpcMetricsTagProvider tagProvider;
+
+  public TaggingMetricCollectingServerInterceptor(
+      MeterRegistry registry, GrpcMetricsTagProvider tagProvider) {
+    super(registry);
+    this.tagProvider = tagProvider;
+  }
+
+  public TaggingMetricCollectingServerInterceptor(
+      MeterRegistry registry,
+      UnaryOperator<Counter.Builder> counterCustomizer,
+      UnaryOperator<Timer.Builder> timerCustomizer,
+      GrpcMetricsTagProvider tagProvider,
+      Status.Code... eagerInitializedCodes) {
+    super(registry, counterCustomizer, timerCustomizer, eagerInitializedCodes);
+    this.tagProvider = tagProvider;
+  }
+
+  @Override
+  public <Q, A> ServerCall.Listener<Q> interceptCall(
+      ServerCall<Q, A> call, Metadata requestHeaders, ServerCallHandler<Q, A> next) {
+    // only first line changed from the super impl
+    final MetricSet metrics = metricsFor(call);
+    final Consumer<Status.Code> responseStatusTiming =
+        metrics.newProcessingDurationTiming(this.registry);
+
+    final MetricCollectingServerCall<Q, A> monitoringCall =
+        new MetricCollectingServerCall<>(call, metrics.getResponseCounter());
+
+    return new MetricCollectingServerCallListener<>(
+        next.startCall(monitoringCall, requestHeaders),
+        metrics.getRequestCounter(),
+        monitoringCall::getResponseCode,
+        responseStatusTiming);
+  }
+
+  private MetricSet metricsFor(ServerCall<?, ?> call) {
+    MethodDescriptor<?, ?> method = call.getMethodDescriptor();
+    Tags callTags = tagProvider.getCallTags(call);
+
+    // check if there is any extension, if not return super
+    // otherwise use our implementation
+    if (Tags.empty().equals(callTags)) {
+      return super.metricsFor(method);
+    } else {
+      return new MetricSet(
```

Comment:
It looks like micrometer tries to cache the `MetricSet` per method if it can: https://github.com/micrometer-metrics/micrometer/blob/v1.7.3/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/grpc/AbstractMetricCollectingInterceptor.java#L164

Is that something we should try to do (using tags/method)? Or is this not a big deal performance-wise?

Reply:
Yea they are caching as there are no extra tags per call.. And they are able to cache per method.. In our case you can not cache, as tags can change on each call..

However, we do use the caching if the tag provider returns empty tags.. So in case of Noop tag provider we will use original impl.. Note that this works because the amount of tags returned by the tag provider have always to be same size..